openquake/hazardlib/shakemap/conditioned_gmfs.py

```diff
@@ -0,0 +1,988 @@
+        """
+        :param gsim: GSIM used to compute mean_stds
+        :param num_events: the number of seismic events
+        :param mean_covs: array of shape (4, M, N)
+        :returns:
+            a 32 bit array of shape (num_imts, num_sites, num_events) and
+            two arrays with shape (num_imts, num_events): sig for tau
+            and eps for the random part
+        """
+        M = len(self.imts)
+        num_sids = mean_covs[0][self.imts[0].string].size
+        result = numpy.zeros((M, num_sids, num_events), F32)
+        sig = numpy.zeros((M, num_events), F32)  # same for all events
+        eps = numpy.zeros((M, num_events), F32)  # not the same
+        numpy.random.seed(self.seed)
+
+        for m, imt in enumerate(self.imts):
+            mu_Y_yD = mean_covs[0][imt.string]
+            # cov_Y_Y_yD = mean_covs[1][imt.string]
+            cov_WY_WY_wD = mean_covs[2][imt.string]
+            cov_BY_BY_yD = mean_covs[3][imt.string]
+            try:
+                result[m], sig[m], eps[m] = self._compute(
+                    mu_Y_yD, cov_WY_WY_wD, cov_BY_BY_yD, imt, num_events, rng
+                )
+            except Exception as exc:
+                raise RuntimeError(
+                    "(%s, %s, source_id=%r) %s: %s"
+                    % (gsim, imt, decode(self.source_id), exc.__class__.__name__, exc)
+                ).with_traceback(exc.__traceback__)
+        if self.amplifier:
+            self.amplifier.amplify_gmfs(self.ctx.ampcode, result, self.imts, self.seed)
+        return result, sig, eps
+
+    def _compute(self, mu_Y, cov_WY_WY, cov_BY_BY, imt, num_events, rng):
+        if self.cmaker.truncation_level == 0:
+            gmf = exp(mu_Y, imt)
+            gmf = gmf.repeat(num_events, axis=1)
+            inter_sig = 0
+            inter_eps = [numpy.zeros(num_events)]
+        else:
+            cov_Y_Y = cov_WY_WY + cov_BY_BY
+            # cov_Y_Y_near = cov_nearest(
+            #     cov_Y_Y, method="clipped", threshold=1e-15, n_fact=100, return_all=False
+            # )
+            # cov_Y_Y_near = numpy.nan_to_num(cov_Y_Y_near).clip(min=0)
+            gmf = exp(
+                rng.multivariate_normal(
+                    mu_Y.flatten(),
+                    cov_Y_Y,
+                    size=num_events,
+                    check_valid="warn",
+                    tol=1e-5,
+                    method="eigh",
+                ),
+                imt,
+            ).T
+            inter_sig = 0
+            inter_eps = 0
+        return gmf, inter_sig, inter_eps  # shapes (N, E), 1, E
+
+
+def get_conditioned_mean_and_covariance(
+    rupture,
+    gmm,
+    station_sitecol,
+    station_data,
+    observed_imt_strs,
+    target_sitecol,
+    target_imts,
+    spatial_correl,
+    cross_correl_between,
+    cross_correl_within,
+    maximum_distance,
+):
+    gmm_name = gmm.__class__.__name__
+    if gmm_name == "ModifiableGMPE":
```

Comment:
Perhaps netter `if hasattr(gmm, 'gmpe')`

Reply:
Updated to `if hasattr(gmm, "gmpe")`